File: unbake/utils/model_file_names.py
# Copyright (C) 2026 syugoji
# SPDX-License-Identifier: GPL-3.0-or-later
#
# ComfyUI-Unbake の一部。著作権の所在を明示してあることが、
# 後から別のライセンスを足せる唯一の担保になる。
"""Strip model file extensions to build comparison keys — the single place.

A 2026-08-16 inventory found the same rule hand-written in **nine places with
five different lists** (``_measurements/normalization_inventory_2026-08-16.md``).
Every one of them was missing ``.sft``, and ``.pt2`` / ``.pkl`` appeared nowhere.
The damage was real: a recipe recording ``ae.sft`` failed to match the installed
``ae.safetensors`` and had its whole prompt rejected.

**Declaring a shared rule is not checking it.** The comment on
``recipeMissingModels.js`` said "same rule as the backend ``normalize_model_name``"
while the two lists differed. Equality between the JavaScript and Python sides is
pinned by ``tests/declared_tests_exist_test.mjs``, which reads both files.

(2026-08-28: this line named a test that did not exist. The same test now also
checks that every test named from source is real.)

This module is about *names*. Whether a file's bytes really are that model is a
different question, answered by ``unbake/utils/model_file_validation.py`` — which
excludes extensions such as ``.onnx`` that carry no container contract here.
"""

from __future__ import annotations

import re
from typing import Any, Tuple

#: ComfyUI's ``folder_paths.supported_pt_extensions`` (measured against 0.27.0)
#: plus ``onnx`` and ``gguf``, which custom nodes use.
#:
#: ``gguf`` was added on 2026-08-31 (``I-20260831-69``).  It was already treated
#: as a model extension by ``models.py`` and ``model_index.py`` -- which had
#: their own copies of this list -- but not here, so the two sides keyed the
#: same file differently: Python indexed ``x.gguf`` as ``x`` while the browser
#: looked it up as ``xgguf``.
#:
#: Order matters: ``pt2`` must precede ``pt`` so the alternation does not split
#: ``model.pt2`` into ``model.`` + ``2``.
MODEL_FILE_EXTENSIONS: Tuple[str, ...] = (
    "safetensors", "sft", "ckpt", "pt2", "pt", "pth", "bin", "pkl", "onnx", "gguf",
)

#: The ComfyUI side on its own (no ``onnx``). Tests compare this to the measured set.
COMFYUI_SUPPORTED_PT_EXTENSIONS: Tuple[str, ...] = (
    "ckpt", "pt", "pt2", "bin", "pth", "safetensors", "pkl", "sft",
)

MODEL_EXTENSION_PATTERN = re.compile(
    r"\.(?:%s)$" % "|".join(MODEL_FILE_EXTENSIONS), re.IGNORECASE
)

_NON_ALPHANUMERIC = re.compile(r"[^a-z0-9]+")


def model_basename(value: Any) -> str:
    """Basename after normalising separators; records arrive with both ``\\`` and ``/``."""

    return str(value or "").replace("\\", "/").rsplit("/", 1)[-1]


def strip_model_extension(value: Any) -> str:
    """Drop only a trailing model extension (no basename step)."""

    return MODEL_EXTENSION_PATTERN.sub("", str(value or ""))


def model_stem(value: Any) -> str:
    """Basename with the model extension removed."""

    return strip_model_extension(model_basename(value))


def model_lookup_key(value: Any) -> str:
    """索引を引くための鍵。**フォルダと本体の拡張子を落として小文字にする。**

    ここは長く**3箇所に手で書かれて**いた（``I-20260831-69``）——
    ``models.py`` の ``_stem`` / ``model_index.py`` の ``name_key`` /
    ``web/panel/modelsView.js`` の ``stemOf``。どれも同じことをしていたが、
    落とす拡張子の一覧が**それぞれ違って**いたので、境界で鍵が食い違った。

    **``os.path.splitext`` を使わない**（写しの側が書いていた理由をここへ移す）。
    あれは最後の ``.`` から後ろを落とすので、拡張子の付いていない名前が
    版番号のところで切れる:

        ``ink-style_A3.1_XL``              -> ``ink-style_a3``（``.1_XL`` を拡張子と誤読）
        ``ink-style_A3.1_XL.safetensors``  -> ``ink-style_a3.1_xl``

    左右で茎が食い違うので、**導入済みなのに「入っていない」**になる。
    ``re-mixmain.fp16`` のような「名前の一部」も、一覧に無いので落ちない。
    """

    return model_stem(value).strip().lower()


def compact_model_name(value: Any) -> str:
    """Comparison key that survives spelling differences.

    Recipes spell the same upscaler as ``R-ESRGAN 4x+ Anime6B`` while ComfyUI
    lists it as ``RealESRGAN_x4plus_anime_6B.pth``; only the alphanumerics
    survive both spellings.
    """

    return _NON_ALPHANUMERIC.sub("", model_stem(value).casefold())

File: web/core/modelFileNames.js
/*
 * Copyright (C) 2026 syugoji
 * SPDX-License-Identifier: GPL-3.0-or-later
 *
 * ComfyUI-Unbake の一部。**この行の意味**——著作権の所在を明示してある限り、
 * 後から別のライセンスを足せる。表示が無いまま配ると、それが言いづらくなる。
 */
/**
 * モデル名から拡張子を落として比較キーを作る、**唯一の場所**。
 *
 * 2026-08-16 の棚卸しで、同じ規則が**9箇所に手書きで5通り**あった
 * （`_measurements/normalization_inventory_2026-08-16.md`）。**どれも `.sft` を欠き**、
 * `.pt2` と `.pkl` はどこにも無かった。実害も出ている——`ae.sft` と記録されたレシピが、
 * 導入済みの `ae.safetensors` が在るのに完全一致に失敗して投入ごと拒否されていた。
 *
 * **宣言は検査ではない。** `recipeMissingModels.js` の該当行には
 * 「バックエンドの `normalize_model_name` と同じ規則」と書いてあったが、
 * 実体はバックエンドとも一致していなかった。同値は
 * `tests/declared_tests_exist_test.mjs` が**両言語のファイルを読んで固定する**。
 *
 * （2026-08-28: ここは長い間、**存在しない検査を2本名指ししていた**。
 * 名前が在ると読んだ人はそこで確かめるのをやめるので、無い検査を指すのは
 * 何も書かないより悪い。同じ検査が、名指しの実在も機械で見ている。）
 *
 * ここで扱うのは**名前の正規化**であって、中身が本当にそのモデルかの検査ではない。
 * 後者は `unbake/utils/model_file_validation.py`（先頭バイトの契約）で、
 * `.onnx` のように容器の契約を持たない拡張子を含む点が違う。
 */

/**
 * ComfyUI が**モデルファイルとして受け付ける拡張子**（`folder_paths.supported_pt_extensions`
 * の実測値・ComfyUI 0.27.0）に `onnx`（一部のカスタムノードが使う）を足したもの。
 *
 * 順序に意味がある: `pt2` は `pt` より**前**に置く。後ろに置くと
 * 交替の左優先で `model.pt2` が `model.` + `2` へ割れる。
 */
export const MODEL_FILE_EXTENSIONS = Object.freeze([
    'safetensors', 'sft', 'ckpt', 'pt2', 'pt', 'pth', 'bin', 'pkl', 'onnx', 'gguf',
]);

/** ComfyUI 側の集合（`onnx` を除いたもの）。テストが実測値と突き合わせる。 */
export const COMFYUI_SUPPORTED_PT_EXTENSIONS = Object.freeze([
    'ckpt', 'pt', 'pt2', 'bin', 'pth', 'safetensors', 'pkl', 'sft',
]);

export const MODEL_EXTENSION_PATTERN = new RegExp(
    `\\.(?:${MODEL_FILE_EXTENSIONS.join('|')})$`, 'i');

/** パス区切りを揃えたうえでの basename。記録は `\` と `/` の両方で来る。 */
export function modelBasename(value) {
    return String(value ?? '').replaceAll('\\', '/').split('/').pop() || '';
}

/** 末尾の拡張子だけを落とす（basename 化はしない）。 */
export function stripModelExtension(value) {
    return String(value ?? '').replace(MODEL_EXTENSION_PATTERN, '');
}

/** basename にして拡張子を落とした名前。A1111 の記録は拡張子を持たない。 */
export function modelStem(value) {
    return stripModelExtension(modelBasename(value));
}

/**
 * 名前の末尾に括弧で埋め込まれた **SHA-10** を取り出す（2026-08-29 実機で確定）。
 *
 * 実機で観測した形（記録 `civitai_128383826` の土台）:
 *
 *     Illustrious/aMixIllustrious_aMix(B199B92EE9).safetensors
 *
 * 手元にはこれが `aMixIllustrious_aMix.safetensors` として入っており、
 * 導入済み索引の `checkpoints.bySha10` には `"b199b92ee9"` の項が在った
 * ——**括弧の中身は、その索引が持っているハッシュそのもの**である。
 *
 * **括弧を外した名前で当てにいかない。** それは推測で、同じ名前の別の版を
 * 掴み得る。ハッシュなら索引に当たった時点で**バイト同一だと確かめられる**。
 *
 * @returns {string|null} 小文字10桁。読めなければ `null`。
 */
export function hashFromModelName(value) {
    const stem = modelStem(value);
    const found = /\(([0-9a-f]{10})\)$/i.exec(stem);
    return found ? found[1].toLowerCase() : null;
}

/**
 * 別名照合用のキー。綴りの揺れ（`R-ESRGAN 4x+ Anime6B` 対
 * `RealESRGAN_x4plus_anime_6B.pth`）を越えるため、英数字だけを残す。
 */
/**
 * 索引を引くための鍵。**フォルダと本体の拡張子を落として小文字にする。**
 *
 * `unbake/utils/model_file_names.py` の `model_lookup_key` と同じ規則
 * （`I-20260831-69`）。ここは長く `web/panel/modelsView.js` の `stemOf` と
 * `models.py` / `model_index.py` に手で書かれており、**落とす拡張子の一覧が
 * それぞれ違って**いたので境界で鍵が食い違った。
 *
 * **最後の `.` から後ろを落とさない**——拡張子の付いていない名前が版番号の
 * ところで切れる（実データの `ink-style_A3.1_XL` → `ink-style_a3`）。
 */
export function modelLookupKey(value) {
    return modelStem(value).trim().toLowerCase();
}

export function compactModelName(value) {
    return modelStem(value).replace(/[^a-z0-9]+/gi, '').toLowerCase();
}
